### PR TITLE
Render the git graph on the server

### PR DESCRIPTION
**Pull request by zeripath**
Rendering the git graph on the server means that we can properly track flows and switch from the Canvas implementation to a SVG implementation.

* This implementation provides a 16 limited color selection
* The uniqued color numbers are also provided
* And there is also a monochrome version
* In addition is a hover highlight that allows users to highlight commits on the same flow.

Closes #12209

Signed-off-by: Andrew Thornton <art27@cantab.net>

# Screenshots
## Color
Screenshot from 2020-07-27 18-52-37
Screenshot from 2020-07-27 21-01-30
## Monochrome
Screenshot from 2020-07-26 21-44-16
Screenshot from 2020-07-26 21-43-11

### CODE REVIEW

**Comment by zeripath** [comment]
@silverwind I used the colors provided in https://github.com/go-gitea/gitea/issues/10217#issuecomment-657000790 but I think they don't quite match up nicely as highlights of each other - should be easily fixable though.

**Comment by silverwind** [comment]
Issue is that 16 colors of the same lightness is too much to get proper distinction. Some hand-picked ones may be better where you can also variate the lightness on them.

**Comment by silverwind** [comment]
Based on those screenshots: The header could use a bit more horizontal padding so it aligns with the content below. 

Not sure I like the buttons, maybe try outline variants? Could probably also add some text to them too.

File: templates/repo/graph.tmpl [hunk 1]
```diff
@@ -2,21 +2,38 @@
...
+			</div></h1>
 			<div id="rel-container">
-				<canvas id="graph-canvas">
-					<ul id="graph-raw-list">
-						{{ range .Graph }}
-							<li><span class="node-relation">{{ .GraphAcii -}}</span></li>
-						{{ end }}
-					</ul>
-				</canvas>
+				<svg xmlns="http://www.w3.org/2000/svg" viewbox="{{Mul .Graph.MinColumn 5}} {{Mul .Graph.MinRow 10}} {{Add (Mul .Graph.Width 5) 5}} {{Mul .Graph.Height 10}}" width="{{Add (Mul .Graph.Width 10) 10}}px">
```
Comment (silverwind) [inline templates/repo/graph.tmpl]:
```suggestion
				<svg viewbox="{{Mul .Graph.MinColumn 5}} {{Mul .Graph.MinRow 10}} {{Add (Mul .Graph.Width 5) 5}} {{Mul .Graph.Height 10}}" width="{{Add (Mul .Graph.Width 10) 10}}px">
```

Not necessary for inline svg

**Comment by silverwind** [comment]
Thought again about colors. Current ones are probably fine, but I'd but the green one first.

File: web_src/js/features/gitgraph.js [hunk 2]
```diff
@@ -1,568 +1,87 @@
...
-    } else if (
-      column + 1 < previousRow.length &&
-      previousRow[column + 1] === '\\'
-    ) {
-      flow = this.takePreviousFlow(column + 1);
+export default async function initGitGraph() {
+  const graphContainer = document.getElementById('git-graph-container');
+  if (!graphContainer) return;
+
+  $('#flow-color-monochrome').click(() => {
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js]:
```suggestion
  $('#flow-color-monochrome').on('click', () => {
```

File: web_src/js/features/gitgraph.js [hunk 3]
```diff
@@ -1,568 +1,87 @@
...
+      const href = $(that).attr('href');
+      if (!href) return;
+      const url = new URL(href, window.location);
+      const params = url.searchParams;
+      params.set('mode', 'monochrome');
+      url.search = `?${params.toString()}`;
+      $(that).attr('href', url.href);
+    });
+  });
+  $('#flow-color-colored').click(() => {
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js]:
```suggestion
  $('#flow-color-colored').on('click', () => {
```

File: web_src/js/features/gitgraph.js [hunk 4]
```diff
@@ -1,568 +1,87 @@
...
-    unitSize: 20,
-    lineWidth: 3,
-    nodeRadius: 4
-  };
-
-
-  const gitGraph = new GitGraph(graphCanvas, graphList, config);
-  gitGraph.draw();
-  graphCanvas.closest('#git-graph-container').classList.add('in');
+  $('#git-graph-container #rev-list li').hover(
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js:49]:
```suggestion
  $('#git-graph-container #rev-list li').on('hover', () => {
```

File: web_src/js/features/gitgraph.js [hunk 5]
```diff
@@ -1,568 +1,87 @@
...
+      $(`#rev-list li[data-flow='${flow}']`).addClass('highlight');
+    },
+    (e) => {
+      const flow = $(e.currentTarget).data('flow');
+      if (flow === 0) return;
+      $(`#flow-${flow}`).removeClass('highlight');
+      $(e.currentTarget).removeClass('hover');
+      $(`#rev-list li[data-flow='${flow}']`).removeClass('highlight');
+    },
+  );
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js]: I guess I would prefer delegated `on` handlers over `.hover`:

````js
$("#git-graph-container").on('mouseenter', '#rev-list li', (e) => {
    // stuff
});
$("#git-graph-container").on('mouseleave', '#rev-list li', (e) => {
    // stuff
});

**Comment by silverwind** [comment]
Some suggestions in https://github.com/silverwind/gitea/commit/a4a59d987b66e0616c39bb22733b603bf891ef42 if you care to pull that in.

**Comment by silverwind** [comment]
I think you can delete `web_src/less/vendor` and the import in `web_src/less/index.less` with this commit.

**Comment by silverwind** [comment]
One last suggestion: Put all git graph related styles in `web_src/less/features/gitgraph.less` and then do a `@import "./features/gitgraph.less";` in `index.less`. We have to modularize the less more, this is a good opportunity.

**Review by silverwind — APPROVED** [review]
Nice work

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
conflicted.

**Comment by zeripath** [comment]
make lg-tm work

**Pull request merged**